### ISSUE TITLE
OpenSSL speed output order issue when redirected

### ISSUE THREAD

**Issue by patrikaxelsson** [jens-maus/amissl#58]
When running for the speed test of the OpenSSL command, it works fine in the shell:
![image](https://user-images.githubusercontent.com/12885607/134803391-0d2410cd-b975-463a-8ad6-2798fe0235d3.png)

However, if its output is redirected to a file like `AmiSSL:OpenSSL speed -evp chacha20-poly1305 >ram:output.txt`, the result part of the 16384 size blocks line appears at the last line instead:
```
10.Ram Disk:> type ram:output.txt 
First we calculate the approximate speed ...
Doing chacha20-poly1305 20480 times on 16 size blocks: 81920 chacha20-poly1305's in 37.08s
Doing chacha20-poly1305 20480 times on 64 size blocks: 20480 chacha20-poly1305's in 27.82s
Doing chacha20-poly1305 20480 times on 256 size blocks: 5120 chacha20-poly1305's in 25.82s
Doing chacha20-poly1305 20480 times on 1024 size blocks: 1280 chacha20-poly1305's in 25.34s
Doing chacha20-poly1305 20480 times on 8192 size blocks: 160 chacha20-poly1305's in 25.16s
Doing chacha20-poly1305 20480 times on 16384 size blocks: OpenSSL 1.1.1l  24 Aug 2021
built on: Wed Aug 25 13:06:38 2021 UTC
options:bn(64,32) md2(char) rc4(char) des(long) aes(partial) idea(int) blowfish(ptr) 
compiler: m68k-amigaos-gcc -resident32 -mcrt=clib2 -m68020-40 -msoft-float -W -Wall -Wno-unused -O1 -I../../libcmt -DAMISSL_COMPILE -D__NO_NET_API -DB_ENDIAN -DOPENSSL_SYS_AMIGA -DOPENSSL_NO_STDIO -D__C_MACROS__ -DNO_INLINE_VARARGS -D__amigaos3__ -DNDEBUG -I../../include -I../../include/netinclude
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
chacha20-poly1305       35.35k       47.11k       50.76k       51.73k       52.10k       52.10k
80 chacha20-poly1305's in 25.16s
```

Tested this on both 3.1 and 3.2 with the same result.

**Comment by Futaura** [comment]
Technically, the first 7 lines shouldn't be in the output file. Those are sent to stderr, whilst the others are sent to stdout. Unfortunately, unlike OS4, OS3 doesn't yet have a separate error stream, so the stdout output stream is used for stderr too. So, there are probably other issues like this where the streams may get flushed out of order.

**Comment by patrikaxelsson** [comment]
3.2 can redirect stderr to another file, however it does in contrast to OS4 keep compatibility with OS3 and redirect both stdout and stderr with >.

This doesn't really have anything to do with this issue, but as this was brought up I feel an obligation to mention that changing behaviour of > in OS4, makes is unnecessarily convoluted to get the same script behaviour in OS3 and OS4, without resorting to destroying script readability with lots of if/else to handle things differently for OS4.

To the issue at hand, is it something that could be solved in a reasonable way, like flushing stderr after the first 7 lines instead of at the end?

**Comment by Futaura** [comment]
If it is an isolated case, maybe. It is not worth the effort to fix this up if there are multiple occurrences in other areas of the app too. However, looking at https://github.com/jens-maus/amissl/blob/38d1bc45e52272e9ea6917dabe418f070af03399/openssl/apps/speed.c#L3428
it looks like it should be flushing anyway the issue is not in the speed code, as it works correctly when not redirecting. I'm thinking this could be an issue with OS3 itself - I mean, when redirecting the output, stderr is going to be a non-interactive stream, and OS3 has probably not set this as line buffered, so it is not ideal if stdout and stderr are both merged when redirecting. We are using buffered i/o in dos.library, so perhaps the solution is to use `SetVBuf()` in https://github.com/jens-maus/amissl/blob/38d1bc45e52272e9ea6917dabe418f070af03399/src/libstubs.c#L19 to make sure we use line buffering instead. I shall give it a try later.

Sorry, I got mixed up with OS3 having the lack of OS4's ErrorOutput() function, but we get this from pr_CES on OS3 anyway which was working since OS3.9. As you probably know, you could use `>ram:output.txt *>*` to mimic OS4's default, which should work the same on both. I forgot to try on OS3, but from what I can gather `>ram:output.txt *>>ram:output.txt` should write stdout and stderr to the same file, but on OS4 stderr seems to then go to NIL:!

**Comment by patrikaxelsson** [comment]
Will be interesting to hear about the result for SetVBuf(). If it is a bug in OS3, it could be reported and fixed in 3.2+.

I did not know about *>* and it does work on OS3.2, cheers :). To make life easier when scripting though, I would need something which redirects both stdout and stderr to NIL: which works in both OS4 and OS3.1.

A classic example is when you attempt to delete a file which might not exist and you do not want to see a stray error message if it doesn't exit.

In OS3.1, it is "delete mightexist >NIL:", in OS4 it is "delete mightexist >NIL: *>NIL:". The OS4 variant brings up "Please insert volume *>NIL in any drive" on OS3.1.

**Comment by Futaura** [comment]
Unfortunately my SetVBuf() idea was flawed, because all file handles open in BUF_LINE mode by default. However, automatic flushing on a '\n' only happens in DOS when the file handle is interactive - that is why it is ok when output is not redirected to a file. Same on OS3 and OS4, and obviously it makes perfect sense for it to be done this way.

So, we arrive back at the way OS3 merges stdout and stderr when redirecting to a file - I'm not sure how it does so, but it can't be using the same file handle for stdout and stderr, otherwise it would be impossible for the flushing of both to become out of sync the way it does currently. I guess it must be doing something funky, but really it should be flushing before switching between stdout and stderr - it is not something that applications should need to take care of.

As I say, whilst I could add a specific flush for the speed command, I'm sure there are many other instances in the OpenSSL command where this will also need to be done. Is there somewhere official to report bugs for OS3.2?

**Comment by Futaura** [comment]
Hmmm... actually, there must be something else going on that I've not figured out. Running here with the OS3.9 shell, pr_CES is NULL, so the OpenSSL tool will use Output() for both stdout and stderr. Therefore, since the prints to stdout and stderr are to the exact same handle, there should be no issue with not flushing stderr (because it is really stdout). I've not tried it on OS3.2. Something doesn't make sense.

I've made a simple bit of code for testing, which can be compiled for OS3 or OS4: [shelloutput.zip](https://github.com/jens-maus/amissl/files/7253270/shelloutput.zip)

For OS4, it seems `>ram:out.txt *>>ram:out.txt` is equivalent to `>ram:out.txt *><`, which do kind of work, but the flushing is out of sync in a way that doesn't appear to make any sense either.

**Comment by patrikaxelsson** [comment]
3.2:
```
8.Temp:shelloutput> version full 
Kickstart 47.7, Workbench 47.1 (2020-02-09)
8.Temp:shelloutput> shelloutput 
Output() => 01CEB1F1
ErrorOutput() => 00000000
Starting...
1
2
3
8.Temp:shelloutput> shelloutput >ram:out.txt 
8.Temp:shelloutput> type ram:out.txt 
Output() => 01C30DF7
ErrorOutput() => 00000000
Starting...
1
2
3
8.Temp:shelloutput> shelloutput >ram:out.txt *>>ram:out.txt 
shelloutput: unable to open redirection file
8.Temp:shelloutput> type ram:out.txt 
8.Temp:shelloutput> shelloutput >ram:out.txt *>ram:out2.txt 
8.Temp:shelloutput> type ram:out.txt 
Output() => 01C30DF7
ErrorOutput() => 01C30E2D
Starting...
3
8.Temp:shelloutput> type ram:out2.txt 
1
2
8.Temp:shelloutput> shelloutput >ram:out.txt *>< 
8.Temp:shelloutput> type ram:out.txt 
Output() => 01C30DF7
ErrorOutput() => 01C30DF7
Starting...
1
2
3
```
`>ram:out.txt *>>ram:out.txt` does not work on 3.2, like it tries to open the same file twice. I have seen this behaviour before.

3.1:
```
10.Temp:shelloutput> version full
Kickstart 40.68, Workbench 40.42 (1994-02-18)
10.Temp:shelloutput> shelloutput
Output() => 01C69E99
ErrorOutput() => 00000000
Starting...
1
2
3
10.Temp:shelloutput> shelloutput >ram:out.txt
10.Temp:shelloutput> type ram:out.txt
Output() => 01C3F0A3
ErrorOutput() => 00000000
Starting...
1
2
3
10.Temp:shelloutput> shelloutput >ram:out.txt *>>ram:out.txt
10.Temp:shelloutput> type ram:out.txt 
Output() => 01C3F0A3
ErrorOutput() => 00000000
Starting...
1
2
3
10.Temp:shelloutput> shelloutput >ram:out.txt *>ram:out2.txt
10.Temp:shelloutput> type ram:out.txt 
Output() => 01C3F0A3
ErrorOutput() => 00000000
Starting...
1
2
3
10.Temp:shelloutput> type ram:out2.txt
TYPE can't open ram:out2.txt
object not found
10.Temp:shelloutput> shelloutput >ram:out.txt *><
10.Temp:shelloutput> type ram:out.txt
Output() => 01C6CA51
ErrorOutput() => 00000000
Starting...
1
2
3
```

Funny detail with 3.1 - `>ram:out.txt *>ram:out.txt` appears to work work on 3.1, but it does not really - as you can see it doesn't create the file and its because its treated as just another word on the command line:
```
10.Temp:shelloutput> echo *>NIL:
*>NIL:
```

Thats why `delete *>NIL:` on 3.1 prompts you to insert drive `*>NIL`.

I see no ordering issues with your shelloutput though, am I supposed to?

Regarding official place to report 3.2 bugs, you can post on the eab forum, they read there frequently.

**Comment by Futaura** [comment]
Nope, looks as expected - just wanted to make sure your system is working the same as mine. This proves it is not really an OS issue.

I think I've pinpointed the real cause: the speed output uses a mixture of printf() for stdout and BIO_printf() for stderr. That becomes a problem on OS3 when redirecting both to the same file (stderr and stdout are the same stream), because printf is used from clib2, which buffers internally and uses unbuffered dos.library writes, whereas BIO_printf() uses buffered dos.library writes directly. Hence the clib2 buffering clashes with the dos.library buffering.

Incidentally, I note AmiSSL v3 had the same issue. Unfortunately, a fix is not straightforward. It is a pity the OpenSSL team mix their usage like this, as in other modules of the app, they use BIO_printf on stdout. There is no need for them to be using printf.

**Comment by patrikaxelsson** [comment]
Could a solution be to use clib buffered output in BIO_printf? Performance wise, it would most likely be a positive thing as dos.library buffered output is extremely slow unless you are using 3.2, where it has been fixed.

**Comment by Futaura** [comment]
Unfortunately not. Being part of the AmiSSL library, BIO_printf has to use native Amiga I/O, as do all the other related functions. They have to use DOS natively, because obviously applications will be built against many different C link libraries, and the FILE structure will be different in them all, so AmiSSL cannot handle these. That is why the stdio variants of OpenSSL functions are not available at all in AmiSSL.

A solution in this case is to map all the printf calls in the OpenSSL app to BIO_printf - I've tested this and it solves the problem. It does set up the BIO streams right at the start of main(), so printf is never called before then. Just need to implement it in the most elegant way now. In some ways it is weird they've even re-implemented the printf family in OpenSSL, but even weirder that the OpenSSL app doesn't use them exclusively - it doesn't need to use stdio. Or they could have used stdio exclusively and not bothered with the BIO stuff. Oh well, as I say, OpenSSL is known for its inconsistency and unnecessary duplication of code :-(.

**Comment by patrikaxelsson** [comment]
Good work finding the issue!

**Comment by Futaura** [comment]
I've added a workaround for OS4 too (I guess it could apply to OS3 at some point also), where the OpenSSL tool will now essentially fallback to operating like OS3 does - that is, if redirecting both stdout and stderr to the same file, we now use the same file handle for stdout and stderr (as before, stdout and stderr were different file handles linked to the exact same file). Maybe there will be a fix for this in OS4 later (I've been discussing it with Colin), but this seems like the best idea for now.

This means that redirecting OpenSSL output to a file should work more consistently on both OS3 (well, 3.2 at least) and OS4 now, where you can use either `> out.txt *><` or `> out.txt *>> out.txt` to redirect both stdout and stderr to the same file.

**Comment by Futaura** [comment]
Remapping `printf()` to `BIO_printf()` was a nice clean fix. However, on further inspection, they are also using `fprintf(stdout` and `fprintf(stderr` - sigh! I don't think a suitable macro can be made to remap those to `BIO_printf()`. So, I'm now working on further fix which maps `printf`, `fprintf`, `BIO_puts` and `BIO_printf` to a new custom routine, which ultimately calls `BIO_vprintf`. One advantage to that is I can add Ctrl-C checking to this function, so the OpenSSL app should respond to a Ctrl-C pretty much everywhere.

**Comment by Futaura** [comment]
@patrikaxelsson Before I commit the latest changes, would you mind testing the latest build? I've included the latest OS3 OpenSSL command in [assl_test.zip](https://github.com/jens-maus/amissl/files/7271391/assl_test.zip) - I've also put the 68020-40 library in there too, which has the rand changes. In summary, this should fix the redirection problems, adds Ctrl-C checking (during writes to output, like stdio) and includes the change back to UNIT_MICROHZ (#57). Now I don't have to reboot my Amiga when I accidentally type `openssl speed` and forget to add the options!

**Comment by patrikaxelsson** [comment]
First the rand changes.

Before with 4.10:
```
8.Ram Disk:> version full 
Kickstart 47.7, Workbench 47.1 (2020-02-09)
8.Ram Disk:> pt:InitAmiSSLTest 
OpenLibrary(amisslmaster.library, 4): 0.150327s
InitAmiSSLMaster(16, 0):              0.000067s
OpenAmiSSL():                         1.914635s
InitAmiSSL(0)                         0.000239s
OPENSSL_init_ssl(0020000E, 0):        1.753169s
GenerateRandomSeed(073CEF04, 128):    0.001962s
RAND_seed(073CEF04, 128):             1.909714s
8.Ram Disk:> pt:InitAmiSSLTest 
OpenLibrary(amisslmaster.library, 4): 0.001313s
InitAmiSSLMaster(16, 0):              0.000069s
OpenAmiSSL():                         0.022016s
InitAmiSSL(0)                         0.000239s
OPENSSL_init_ssl(0020000E, 0):        1.824256s
GenerateRandomSeed(073CF304, 128):    0.001963s
RAND_seed(073CF304, 128):             1.916134s
8.Ram Disk:> version amissl_v111l.library full 
amissl_v111l.library 4.10 (2021-08-25)
os3-68020 version
```

With the version in assl_test.zip:
```
8.Ram Disk:> version full 
Kickstart 47.7, Workbench 47.1 (2020-02-09)
8.Ram Disk:> pt:InitAmiSSLTest 
OpenLibrary(amisslmaster.library, 4): 0.152764s
InitAmiSSLMaster(16, 0):              0.000066s
OpenAmiSSL():                         1.882525s
InitAmiSSL(0)                         0.000238s
OPENSSL_init_ssl(0020000E, 0):        1.721912s
GenerateRandomSeed(073D03D4, 128):    0.001965s
RAND_seed(073D03D4, 128):             0.098195s
8.Ram Disk:> pt:InitAmiSSLTest 
OpenLibrary(amisslmaster.library, 4): 0.001391s
InitAmiSSLMaster(16, 0):              0.000063s
OpenAmiSSL():                         0.025106s
InitAmiSSL(0)                         0.000239s
OPENSSL_init_ssl(0020000E, 0):        1.799022s
GenerateRandomSeed(073D069C, 128):    0.001965s
RAND_seed(073D069C, 128):             0.097816s
8.Ram Disk:> version full amissl_v111l.library 
amissl_v111l.library 4.11 (2021-08-25)
os3-68020 version
```

RAND_seed() is now ~19.5 times faster - a healthy improvement to say the least :).

So to OpenSSL speed redirection.

For 3.2:
```
8.Ram Disk:> version full 
Kickstart 47.7, Workbench 47.1 (2020-02-09)
8.Ram Disk:> version full Temp:assl_test/AmiSSL/C/AmigaOS3/OpenSSL 
OpenSSL 1.1.1l  24 Aug 2021
8.Ram Disk:> Temp:assl_test/AmiSSL/C/AmigaOS3/OpenSSL speed -evp chacha20-poly1305 >ram:output.txt
8.Ram Disk:> type ram:output.txt  
First we calculate the approximate speed ...
Doing chacha20-poly1305 20480 times on 16 size blocks: 81920 chacha20-poly1305's in 38.42s
Doing chacha20-poly1305 20480 times on 64 size blocks: 20480 chacha20-poly1305's in 28.90s
Doing chacha20-poly1305 20480 times on 256 size blocks: 5120 chacha20-poly1305's in 26.38s
Doing chacha20-poly1305 20480 times on 1024 size blocks: 1280 chacha20-poly1305's in 25.76s
Doing chacha20-poly1305 20480 times on 8192 size blocks: 160 chacha20-poly1305's in 25.60s
Doing chacha20-poly1305 20480 times on 16384 size blocks: 80 chacha20-poly1305's in 25.60s
OpenSSL 1.1.1l  24 Aug 2021
built on: Sat Oct  2 11:30:44 2021 UTC
options:bn(64,32) md2(char) rc4(char) des(long) aes(partial) idea(int) blowfish(ptr) 
compiler: m68k-amigaos-gcc -resident32 -mcrt=clib2 -m68020-40 -msoft-float -W -Wall -Wno-unused -O1 -I../../libcmt -DAMISSL_COMPILE -D__NO_NET_API -DB_ENDIAN -DOPENSSL_SYS_AMIGA -DOPENSSL_NO_STDIO -D__C_MACROS__ -DNO_INLINE_VARARGS -D__amigaos3__ -DNDEBUG -I../../include -I../../include/netinclude
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
chacha20-poly1305       34.12k       45.35k       49.69k       50.88k       51.20k       51.20k
```

For 3.1:
```
11.Ram Disk:> version full
Kickstart 40.68, Workbench 40.42 (1994-02-18)
11.Ram Disk:> version full Temp:assl_test/AmiSSL/C/AmigaOS3/OpenSSL
OpenSSL 1.1.1l  24 Aug 2021
11.Ram Disk:> Temp:assl_test/AmiSSL/C/AmigaOS3/OpenSSL speed -evp chacha20-poly1305 >ram:output.txt
11.Ram Disk:> type ram:output.txt 
First we calculate the approximate speed ...
Doing chacha20-poly1305 20480 times on 16 size blocks: 81920 chacha20-poly1305's in 37.52s
Doing chacha20-poly1305 20480 times on 64 size blocks: 20480 chacha20-poly1305's in 27.94s
Doing chacha20-poly1305 20480 times on 256 size blocks: 5120 chacha20-poly1305's in 25.86s
Doing chacha20-poly1305 20480 times on 1024 size blocks: 1280 chacha20-poly1305's in 25.40s
Doing chacha20-poly1305 20480 times on 8192 size blocks: 160 chacha20-poly1305's in 25.26s
Doing chacha20-poly1305 20480 times on 16384 size blocks: 80 chacha20-poly1305's in 25.24s
OpenSSL 1.1.1l  24 Aug 2021
built on: Sat Oct  2 11:30:44 2021 UTC
options:bn(64,32) md2(char) rc4(char) des(long) aes(partial) idea(int) blowfish(ptr) 
compiler: m68k-amigaos-gcc -resident32 -mcrt=clib2 -m68020-40 -msoft-float -W -Wall -Wno-unused -O1 -I../../libcmt -DAMISSL_COMPILE -D__NO_NET_API -DB_ENDIAN -DOPENSSL_SYS_AMIGA -DOPENSSL_NO_STDIO -D__C_MACROS__ -DNO_INLINE_VARARGS -D__amigaos3__ -DNDEBUG -I../../include -I../../include/netinclude
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
chacha20-poly1305       34.93k       46.91k       50.69k       51.60k       51.89k       51.93k
```

The problems in this issue regarding redirection is solved both in 3.1 and 3.2, but you have gone beyond that, very nice with the complete remap of all stdio functions and the combination with the ctrl-c ability on that, good work!
